src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -1424,6 +1427,384 @@ g.test('inputs,sample_mask')
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: framebuffer.createView(),
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bg);
+    pass.draw(3, 1, i);
+    pass.end();
+    t.queue.submit([encoder.finish()]);
+
+    const buffer = t.copyWholeTextureToNewBufferSimple(framebuffer, 0);
+    const readback = await t.readGPUBufferRangeTyped(buffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: uintLength,
+      method: 'copy',
+    });
+    const data: Uint32Array = readback.data;
+
+    t.expectOK(checker(data));
+  }
+}
+
+g.test('subgroup_size')
+  .desc('Tests subgroup_size values')
+  .params(u =>
+    u
+      .combine('size', kSizes)
+      .beginSubcases()
+      .combineWithParams([{ format: 'rgba32uint' }] as const)
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(async t => {
+    interface SubgroupLimits extends GPUSupportedLimits {
+      minSubgroupSize: number;
+      maxSubgroupSize: number;
+    }
+    const { minSubgroupSize, maxSubgroupSize } = t.device.limits as SubgroupLimits;
+
+    const fsShader = `
+enable subgroups;
+
+const width = ${t.params.size[0]};
+const height = ${t.params.size[1]};
+
+@group(0) @binding(0) var<storage, read_write> var for_layout : u32;
+
+@fragment
+fn fsMain(
+  _ = for_layout;
+
+  @builtin(position) pos : vec4f,
+  @builtin(subgroup_size) sg_size : u32,
+) -> @location(0) vec4u {
+  let ballot = countOneBits(subgroupBallot(true));
+  let ballotSize = ballot.x + ballot.y + ballot.z + ballot.w;
+
+  // Do all invocations in the subgroup see the same subgroup size?
+  let firstSize = subgroupBroadcast(sg_size, 0);
+  let compareBallot = countOneBits(subgroupBallot(firstSize == sg_size));
+  let compareSize = compareBallot.x + compareBallot.y + compareBallot.z + compareBallot.w;
+  let sameSize = select(0u, 1u, compareSize == ballotSize);
+
+  return vec4u(sg_size, ballotSize, sameSize, 0);
+}`;
+
+    await runSubgroupTest(
+      t,
+      t.params.format,
+      fsShader,
+      t.params.size[0],
+      t.params.size[1],
+      (data: Uint32Array) => {
+        return checkSubgroupSizeConsistency(
+          data,
+          t.params.format,
+          minSubgroupSize,
+          maxSubgroupSize,
+          t.params.size[0],
+          t.params.size[1]
+        );
+      }
+    );
+  });
+
+/**
+ * Checks subgroup_invocation_id value consistency
+ *
+ * Very little uniformity is expected for subgroup_invocation_id.
+ * This function checks that all ids are less than the subgroup size
+ * and no id is repeated.
+ * @param data An array of vec4u that contains (per texel):
+ *             * subgroup_invocation_id
+ *             * ballot size
+ *             * representative subgroup id
```

Comment:
I was slightly confused by "representative subgroup id" because I thought briefly it was a system-provided subgroup ID;  related, that 0 would be a possible value.

Might help to say:  A non-zero ID unique to each subgroup.  ?